worker/credentialvalidator/worker_test.go

```diff
@@ -0,0 +1,165 @@
+// Copyright 2018 Canonical Ltd.
+// Licensed under the LGPLv3, see LICENCE file for details.
+
+package credentialvalidator_test
+
+import (
+	"github.com/juju/errors"
+	"github.com/juju/testing"
+	jc "github.com/juju/testing/checkers"
+	gc "gopkg.in/check.v1"
+	"gopkg.in/juju/names.v2"
+
+	"github.com/juju/juju/api/base"
+	"github.com/juju/juju/worker/credentialvalidator"
+	"github.com/juju/juju/worker/workertest"
+)
+
+type WorkerSuite struct {
```

Comment:
There need to be tests that the worker and watcher are properly stopped.
See other worker packages for examples.

Reply:
Again, all these are similar to model migration flag. All the tests here that get a valid watcher/worker have workertest.CheckKilled etc... Did you look at:
TestModelCredentialErrorWhileRunning l. 64;
TestCredentialChangeToInvalid l.91;
TestCredentialChangeFromInvalid l.110; 
TestModelCredentialReplaced l. 129;
TestNoRelevantCredentialChange l.147.

I have also ran all the tests in other packages that test workers start and stop (like cmd/jujud/agent).

In addition, I have live tested.

Reply:
We often have a simple test like this:

func (s *WorkerSuite) TestStartStop(c *gc.C) {
	w, err := caasoperator.NewWorker(s.config)
	c.Assert(err, jc.ErrorIsNil)
	workertest.CheckAlive(c, w)
	workertest.CleanKill(c, w)
}

Also, there's no check that I can see that the watcher is killed when the worker dies.

Reply:
Don't necessarily assume that what's in migration flag or elsewhere is current best practice. Sadly this stuff tends to evolve and there's a bit of legacy code all around.

Reply:
I do agree with Ian on this one, traditionally workers have a simple test that just has the start and a clean kill.